src/redisearch_rs/rqe_iterators/src/intersection.rs

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! Supporting types for [`Intersection`].
+//!
+//! # TODO (MOD-12717)
+//!
+//! The following features from the C++ implementation are not yet ported:
+//! - `max_slop`: Maximum allowed slop between term positions
+//! - `in_order`: Require terms to appear in order
+//!
+//! Currently, the iterator operates with `max_slop = -1` semantics (no slop validation).
+
+use ffi::t_docId;
+use inverted_index::RSIndexResult;
+
+use crate::{RQEIterator, RQEIteratorError, RQEValidateStatus, SkipToOutcome};
+
+/// Yields documents appearing in ALL child iterators using a merge (AND) algorithm.
+///
+/// Children are sorted by estimated result count (smallest first) to minimize iterations.
+/// A document is only yielded when ALL children have a matching entry for it.
+pub struct Intersection<'index, I> {
+    /// Child iterators, sorted by estimated count (smallest first).
+    children: Vec<I>,
+    /// Last doc_id successfully found in ALL children (returned by `last_doc_id()`).
+    last_doc_id: t_docId,
+    num_expected: usize,
+    is_eof: bool,
+    /// Aggregate result combining children's results, reused to avoid allocations.
+    result: RSIndexResult<'index>,
+}
+
+/// Result of attempting to get all children to agree on a target document ID.
+enum AgreeResult {
+    /// All children have the target doc_id as their current position.
+    Agreed,
+    /// A child skipped past the target to a higher doc_id; consensus must restart from this new ID.
+    Ahead(t_docId),
+    /// A child reached EOF; no more documents can match.
+    Eof,
+}
+
+impl<'index, I> Intersection<'index, I>
+where
+    I: RQEIterator<'index>,
+{
+    /// Creates a new intersection iterator. Children are sorted by estimated count. If `children`
+    /// is empty, returns an iterator immediately at EOF.
+    #[must_use]
+    pub fn new(mut children: Vec<I>) -> Self {
+        children.sort_by_cached_key(|c| c.num_estimated());
+        let Some(num_expected) = children.first().map(|c| c.num_estimated()) else {
+            return Self {
+                children,
+                last_doc_id: 0,
+                num_expected: 0,
+                is_eof: true,
+                result: RSIndexResult::intersect(0),
+            };
+        };
+
+        let result = RSIndexResult::intersect(children.len());
```

Comment:
For consistency with the `return Self { }` above we might want to inline the value of the `result` variable in the below expression.